sstables/trie.cc

```diff
@@ -30,6 +30,155 @@ namespace trie {
 
 using namespace trie;
 
+// Pimpl boilerplate
+bti_trie_sink::bti_trie_sink() = default;
+bti_trie_sink::~bti_trie_sink() = default;
+bti_trie_sink& bti_trie_sink::operator=(bti_trie_sink&&) = default;
+bti_trie_sink::bti_trie_sink(std::unique_ptr<bti_trie_sink_impl> x) : _impl(std::move(x)) {}
+
+static_assert(trie_writer_sink<bti_trie_sink_impl>);
+
+// The job of this class is to take a stream of partition keys and their index payloads
+// (Data.db or Rows.db position, and hash bits for filtering false positives out)
+// and translate it to a minimal trie which still allows for efficient queries.
+//
+// For that, each key in the trie is trimmed to the shortest length which still results
+// in a unique leaf.
+// For example, assume that keys ABCXYZ, ABDXYZ, ACXYZ are added to the trie.
+// A "naive" trie would look like this:
+//     A
+//     B---C
+//     C-D X
+//     X X Y
+//     Y Y Z
+//     Z Z
+//
+// This writer creates:
+//     A
+//     B---C
+//     C-D
+// 
+// Note: the trie-serialization format of partition keys is prefix-free,
+// which guarantees that only leaf nodes carry a payload. 
+// 
+// In practice, Output == trie_serializer.
+// But for testability reasons, we don't hardcode the output type.
+// (So that tests can test just the logic of this writer, without
+// roping in file I/O and the serialization format)
```

Comment:
File I/O can be abstracted via an interface. Is it important to keep the serialization format abstracted?

Reply:
Keeping the serialization format abstracted is important for the lower layer, which tests the behaviour of node packing at page boundaries, etc. To trigger all the interesting logic, the test has to construct nodes of particular sizes, which would be impractical with the target specialization format.

At this level, keeping the serialization format abstracted isn't important. But I also don't see downsides.

As far as compile time hygiene etc. goes, note that this templatization is hidden behind a pimpl.